{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<table style=\"background-color:#F5F5F5;\" width=\"100%\">\n",
    "<tr><td style=\"background-color:#F5F5F5;\"><img src=\"../images/logo.png\" width=\"150\" align='right'/></td></tr>     <tr><td>\n",
    "            <h2><center>Aprendizagem Automática em Engenharia Biomédica</center></h2>\n",
    "            <h3><center>1st Semester - 2024/2025</center></h3>\n",
    "            <h4><center>Universidade Nova de Lisboa - Faculdade de Ciências e Tecnologia</center></h4>\n",
    "</td></tr>\n",
    "    <tr><td><h2><b><center>Lab 3 - Genetic Algorithms</center></b></h2>\n",
    "    <h4><i><b><center>Optimizing Parameters through System Evolution</center></b></i></h4></td></tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Introduction to Genetic Algorithms\n",
    "\n",
    "### 1.1 What Are Genetic Algorithms?\n",
    "Genetic Algorithms (GAs) are a type of optimization algorithm inspired by the process of natural selection in biology. They are useful for solving complex problems by mimicking the principles of evolution, such as __selection__, __mutation__, and __crossover__ (exchange of DNA between a pair of homologous chromosomes). Just like in nature, where species evolve to become better suited to their environments, GAs help us find the best solution to a problem by iteratively improving candidate solutions.\n",
    "\n",
    "### 1.2 Why Use Genetic Algorithms?\n",
    "Genetic algorithms are particularly useful for problems that:\n",
    "\n",
    "* Have a large search space (many possible solutions).\n",
    "* Don’t have a clear mathematical formula for finding the optimal solution.\n",
    "* Involve multiple competing factors (like cost vs. effectiveness).\n",
    "\n",
    "In Biomedical Engineering, GAs can help optimize [medical treatments](https://www.sciencedirect.com/science/article/pii/S0957417421017450), [drug dosages](https://ascpt.onlinelibrary.wiley.com/doi/full/10.1002/psp4.12512), [signal processing](https://www.mdpi.com/2673-7426/3/4/71), and even the design of medical devices like [proshtetics](https://link.springer.com/article/10.1007/s10237-019-01258-7). GAs are well-suited to explore solutions where traditional methods struggle.\n",
    "\n",
    "### 1.3 How Do Genetic Algorithms Work?\n",
    "The process of a GA can be broken down into several key steps:\n",
    "\n",
    "* __Initialization__: Start with a random set of possible solutions, known as the population.\n",
    "* __Fitness Function__: Each solution is evaluated based on how well it solves the problem (using a fitness function).\n",
    "* __Selection__: The best solutions are chosen based on their fitness to move on to the next generation.\n",
    "* __Crossover (Recombination)__: New solutions are created by combining parts of two or more parent solutions.\n",
    "* __Mutation__: Random changes are made to some solutions to introduce variability.\n",
    "* __Termination__: The process is repeated until the algorithm finds a solution that meets the desired criteria or after a set number of iterations.\n",
    "\n",
    "### 1.4 Goals of Today's Class\n",
    "In today’s practical session, you will learn to apply genetic algorithms to two problems. A more generic one that will be used to give you an understanding on how to implement GAs and one related to Biomedical Engineering:\n",
    "* __The Knapsack Problem:__ You will use a GA to optimize the amount of equipment you can fit into your backpack for a camping trip.\n",
    "* __Optimizing Parameters for Biomedical Signal Processing:__ You will use a GA to improve how well a filter cleans up noisy ECG signals.\n",
    "  \n",
    "We will guide you through the steps of setting up the genetic algorithm, coding it in Python, and evaluating the results. By the end of the class, you will have a better understanding of GAs and how they can be used in Biomedical Engineering."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. A Simple Example Of Genetic Algorithms in Use: The Knapsack Problem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we start with the exerercises we are going to look at a simple toy example where GAs can be utilized. This example will help you to get familiarized with how GAs can be implemented in Python. For this toy example we will take a look at the __\"Knapsack Problem\"__. \n",
    "\n",
    "The [__Computerphile__](https://www.youtube.com/@Computerphile/videos) channel on YouTube made a great [<u>video</u>](https://www.youtube.com/watch?v=MacVqujSXWE) about it and we will essentially implement what is being discussed in the video. The channel is a great resource if you want to explore computer science topics. (The youtuber behind that channel has also other channels covering topics such as [Maths](https://www.youtube.com/@numberphile), [Physics](https://www.youtube.com/@sixtysymbols), [Chemistry](https://www.youtube.com/@periodicvideos), and [Astronomy](https://www.youtube.com/@DeepSkyVideos), just in case you want to check them out after the class)\n",
    "\n",
    "### 2.1 Problem Definition\n",
    "\n",
    "Imagine you have a backpack (knapsack) with a limited capacity, and you are given a set of items, each with a certain weight and value (utility points). The goal is to maximize the total value of the items you can fit in the knapsack without exceeding its weight capacity. The challenge is to find the best combination of items to maximize the value, while respecting the knapsack's weigt capacity.\n",
    "\n",
    "With regards to the GA steps defined above, we can formulate the Knapsack Problem as a GA, the following way:\n",
    "\n",
    "* __Initialization__: Start with a population of possible solutions each representing a different combination of items in the knapsack.\n",
    "* __Fitness Function__: Evaluate each solution based on the total value of the items in the knapsack, but penalize solutions that exceed the weight limit.\n",
    "* __Selection__: Select the best solutions (those with higher values and within the weight limit) to move to the next generation.\n",
    "* __Crossover__: Combine solutions to create new ones (for example, taking some items from one solution and others from another).\n",
    "* __Mutation__: Randomly add or remove items to introduce variability.\n",
    "* __Termination__: Repeat until the algorithm finds a good solution (e.g., after a set number of generations or when the solution stops improving).\n",
    "\n",
    "\n",
    "In our toy example we will go an a camping trip. We have __8 items that we can choose to take with us__, however our backpack only has a __maximum weight capacity of 10&nbsp;Kg__. The items with their repsective weights and values (utility points) are shown in the table below\n",
    "\n",
    "| Item              | Weight (kg) | Value (utility points) |\n",
    "|-------------------|-------------|------------------------|\n",
    "| Tent              | 5.0         | 60                     |\n",
    "| Sleeping Bag      | 2.5         | 40                     |\n",
    "| Camping Stove     | 3.0         | 30                     |\n",
    "| Water Bottle      | 1.0         | 10                     |\n",
    "| First Aid Kit     | 1.5         | 20                     |\n",
    "| Flashlight        | 0.5         | 15                     |\n",
    "| Food Supplies     | 4.0         | 50                     |\n",
    "| Jacket            | 1.5         | 25                     |\n",
    "\n",
    "|Maximum Backpack Weight Capacity: | 10 Kg|\n",
    "|----------------------------------|------|\n",
    "\n",
    "### 2.2 Representing the Data as \"Gene Vectors\"\n",
    "In order to solve the Knapsack Problem using a genetic algorithm, we need a way to represent potential solutions (or individuals) as a set of \"genes\" that can be evolved over time. In the case of the Knapsack Problem we can choose to represent (or encode) the potential solution as a binary vector, where each position in the vector represents whether or not a particular item is included in the knapsack.\n",
    "\n",
    "Since we have 8 items for our camping trip, the gene vector will have a length of 8. Each element in the vector will be either:\n",
    "\n",
    "* __1__: The item is included in the knapsack.\n",
    "* __0__: The item is not included in the knapsack.\n",
    "\n",
    "For example, let's consider the following binary vector: [1, 0, 1, 0, 0, 1, 0, 1]\n",
    "\n",
    "This vector represents a possible solution to the problem. In this case:\n",
    "\n",
    "| Item              | Item Number | Vector Position | Item State   | Binary Vector |\n",
    "|-------------------|:-----------:|:---------------:|--------------|:-------------:|\n",
    "| Tent              | 1           | 0               | included     |1              |\n",
    "| Sleeping Bag      | 2           | 1               | not included |0              |\n",
    "| Camping Stove     | 3           | 2               | included     |1              |\n",
    "| Water Bottle      | 4           | 3               | not included |0              |\n",
    "| First Aid Kit     | 5           | 4               | included     |0              |\n",
    "| Flashlight        | 6           | 5               | not included |1              |\n",
    "| Food Supplies     | 7           | 6               | not included |0              |\n",
    "| Jacket            | 8           | 7               | included     |1              |\n",
    "\n",
    "\n",
    "As stated above, the binary vector describing the individual represents __one possible solution__. With the genetic algorithm, we will created multiple individuals and then let these evolve over several generations to find the optimal solution.\n",
    "\n",
    "The usage of binary encoding is a good choice for the Knapsack Problem as in this case an item can only take a binary state. Either the item is included in the backpack or not. Furthermore the chosen encoding has has the following properties:\n",
    "\n",
    "* __Simplicity__: A binary vector is simple to manipulate and clearly shows which items are selected.\n",
    "* __Evolution__: We can easily apply genetic operators like mutation (flipping 0 to 1 or 1 to 0) and crossover (combining parts of two parent solutions) to these binary vectors.\n",
    "* __Evaluation__: Calculating the total weight and value of selected items is straightforward — we simply sum the weights and values of the items corresponding to 1s in the vector.\n",
    "\n",
    "Depending on the problem that you try to solve, other encodings might be more benefical. If you want to find out about other encodings for genetic algorithms you can ask [ChatGPT](https://chatgpt.com/), [Claude](https://claude.ai/login), [Perplexity](https://www.perplexity.ai/), [Elicit](https://elicit.com/welcome), or any other LLM of your choice, or just do a simple Google search. Here is an exemplarly prompt that you could use. You can adapt it to better suit your needs.\n",
    "\n",
    "__Prompt__: _\"What types of encoding are used for genetic algorithms? Please give me for each encoding a general overview, then the corresponding mathematical description, and finally some use cases in which each encoding is used. If suitable, include examples from the field of Biomedical Engineering._\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Implementing the Knapsack Problem in Python: Step-By-Step\n",
    "\n",
    "Now that we have discussed the core concepts, we can start implementing the genetic algorithm using Python. We will do this step-by-step, according to the steps defined in Section 2.1. For each step we will implement a function and then we will finally combine them to a full code example.\n",
    "\n",
    "#### 2.3.1 Import of Necessary Modules, Set Random Seed, and Definition of Knapsack and Item Properties\n",
    "\n",
    "As with any Python code, first we need to import the modules we need. The main module we are going to use today is __numpy__. In accordance with Good Programming Practices we will use typing hints, docstrings (in reStructuredText format - you can of course use another format), and will comment our code. \n",
    "\n",
    "Since we are going to generate some random numbers later in the code, we will set a random seed. This is needed to make this __code reproducible__. If the seed is set, you will always obtain the same result. In the case of our code we will set the seed through numpy as we will use numpy's random module in the code. Whenever you want to have reproducible results in Machine Learning it is beneficial to report the seed that you used to obtain the results (depending on the modules you use, you'll have to use the seed setting function of that specific module). Often times seed = 42, as a reference to the book \"The Hitchhiker's Guide to the Galaxy\", but you can set it to any number you like.\n",
    "\n",
    "Additionally, we will define the knapsack and item properties according to the table above, so that we can use them later in the code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import needed modules for the code\n",
    "import numpy as np\n",
    "from typing import Optional, Tuple, Dict, Union, List\n",
    "import matplotlib.pyplot as plt\n",
    "from tqdm import tqdm  # module for showing neat progress bar when for loops are running\n",
    "\n",
    "# Set the seed for reproducibility\n",
    "random_seed = 42\n",
    "\n",
    "# define knapsack and item properties\n",
    "item_weights = np.array([5.0, 2.5, 3.5, 1.0, 1.5, 0.5, 4.0, 1.5])\n",
    "item_values = np.array([60, 40, 30, 10, 20, 15, 50, 25])\n",
    "max_capacity = 10\n",
    "item_names = [\"tent\", \"sleeping bag\", \"camping stove\", \"water bottle\", \"first aid kit\", \"flashlight\", \"food supplies\", \"jacket\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3.2 Initilalization\n",
    "\n",
    "Now we will initialize our population of possible solutions/individuals. There are two main parameters that we need to set for initialization. These are:\n",
    "\n",
    "* __population size__: the number of individuals (possible solutions) in the population\n",
    "* __num_items__: the number of items that we want to consider in the Knapsack Problem\n",
    "* __seed__: (Optional) seed for the random number generator to ensure reproducibility.\n",
    "\n",
    "For the initialization step, we will simply generate __population_size-many__ random solutions and store them in an array. Remember that each solutions is a binary vector. For that purpose we will use numpy's [__random.randint()__](https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html) function that allows us to generate vectors/arrays that contain integers in a certain range."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def initialize_population(population_size: int, num_items: int, seed: Optional[int] = None) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Initialize a population of binary vectors for the Knapsack Problem.\n",
    "\n",
    "    :param population_size: The number of individuals in the population.\n",
    "    :param num_items: The number of items in the Knapsack Problem.\n",
    "    :param seed: Optional seed for the random number generator to ensure reproducibility.\n",
    "    :return: A 2D NumPy array where each row is a binary vector representing an individual.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Set the seed for reproducibility\n",
    "    if seed is not None:\n",
    "        np.random.seed(seed)\n",
    "    \n",
    "    # Initialize a population with random binary vectors\n",
    "    # passing low=2 will draw numbers out of a uniform distribution using the half open interval [0, 2), meaning it can only draw 0 or 1\n",
    "    population = np.random.randint(low=2, size=(population_size, num_items))\n",
    "    \n",
    "    return population"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just to get an understanding of the function defined above, let's use it with the following parameters. (You can try running the function multiple times with and without the seed to see the difference in output):\n",
    "\n",
    "* __population_size__: 10\n",
    "* __num_items__: len(item_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The initial population is: \n",
      "[[0 1 0 0 0 1 0 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 1 1 0 1 0 1 1]\n",
      " [1 1 1 1 1 1 0 0]\n",
      " [1 1 1 0 1 0 0 0]\n",
      " [0 0 1 1 1 1 1 0]\n",
      " [1 1 0 1 0 1 0 1]\n",
      " [1 0 0 0 0 0 0 0]\n",
      " [0 1 1 0 1 1 1 1]\n",
      " [0 1 0 1 1 1 0 1]]\n"
     ]
    }
   ],
   "source": [
    "# get the number of items (in our example this is 8)\n",
    "num_items = len(item_names)\n",
    "\n",
    "# create an initial population\n",
    "# to run the function without seed, either remove the seed parameter from the line of code or set seed=None\n",
    "initial_population = initialize_population(population_size=10, num_items=num_items, seed=random_seed)\n",
    "\n",
    "# print the result\n",
    "print(\"The initial population is: \\n{}\".format(initial_population))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3.2 Fitness Function\n",
    "\n",
    "Next we will implement the fitness function. Remember that our Knapsack can only carry items up to a specific weight. Thus the fitness function has to fulfill the following steps:\n",
    "\n",
    "For each individual (potential solution) in the __population__:\n",
    "\n",
    "1. calculate the sum of the __weights__ for the chosen items.\n",
    "2. check if the weight is below the weight __capacity__.\n",
    "3. if the weight exceededs the capacity, __return a fitness score of 0__.\n",
    "4. if the weight is below or equal to the weight capacity, __calculate the sum of the values__ (utility points) and return this sum as the fitness score.\n",
    "\n",
    "Therefore, the function needs the following parameters:\n",
    "\n",
    "* __weights__: A 1D array of item weights.\n",
    "* __values__: A 1D array of item values (utility points).\n",
    "* __population__: A 2D array representing the population.\n",
    "* __max_capacity___: The maximum allowable weight the knapsack can carry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_fitness_scores(weights: np.ndarray, values: np.ndarray, population: np.ndarray, max_capacity: int) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Calculate the fitness score of each individual in the population for the Knapsack Problem.\n",
    "\n",
    "    The fitness is the total value of selected items, but if the total weight exceeds the knapsack capacity, \n",
    "    the fitness is penalized (set to 0).\n",
    "\n",
    "    :param weights: A 1D array of item weights.\n",
    "    :param values: A 1D array of item values.\n",
    "    :param population: A 2D array representing the population (each row is a binary vector of item selections).\n",
    "    :param max_capacity: The maximum allowable weight the knapsack can carry.\n",
    "    :return: A 1D array where each entry is the fitness value of the corresponding individual in the population.\n",
    "    \"\"\"\n",
    "\n",
    "    # get the population size\n",
    "    population_size = population.shape[0]\n",
    "    \n",
    "    # initialize an array to store the fitness score of each individual\n",
    "    # for simplicity we create an array of zeros corresponding to the population size\n",
    "    fitness_scores = np.zeros(population_size)\n",
    "\n",
    "    # loop over each individual (potential solution) in the population\n",
    "    for individual_num, individual in enumerate(population):\n",
    "        \n",
    "        # calculate the total weight of the selected items contained in the individual\n",
    "        total_weight = np.sum(individual * weights)\n",
    "\n",
    "        # check if the total_weight is below or equal to the capacity\n",
    "        if total_weight <= max_capacity:\n",
    "\n",
    "            # calculat the total value of the selected items\n",
    "            total_value = np.sum(individual * values)\n",
    "\n",
    "            # assign the total_value as the fitness score\n",
    "            fitness_scores[individual_num] = total_value\n",
    "        \n",
    "        # capacity was exceeded\n",
    "        else:\n",
    "            # penalize by setting fitness to 0\n",
    "            fitness_scores[individual_num] = 0\n",
    "\n",
    "    return fitness_scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's test the function using the small population we generated and the knapsack and item porperties defined above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The calculated fitness scores for all individuals in the population are: [ 55.  90.   0.   0.   0.   0.   0.  60.   0. 110.]\n",
      "The highest value score (110.0) was achieved using the following combination: [0 1 0 1 1 1 0 1]\n",
      "This corresponds to the following items: ['sleeping bag', 'water bottle', 'first aid kit', 'flashlight', 'jacket']\n"
     ]
    }
   ],
   "source": [
    "# calculate the fitness scores for the generated population\n",
    "fitness_scores = calculate_fitness_scores(item_weights, item_values, initial_population, max_capacity)\n",
    "\n",
    "# get the solution that that has the highest score\n",
    "idx_best = np.argmax(fitness_scores)\n",
    "\n",
    "# get the best score\n",
    "best_score = np.max(fitness_scores)\n",
    "\n",
    "# get the best solution in the population\n",
    "best_solution = initial_population[idx_best, :]\n",
    "\n",
    "# get the items that correspond to that solution\n",
    "best_solution_items = [item_names[item_num] for item_num in range(len(item_names)) if best_solution[item_num] == 1]\n",
    "\n",
    "# print the results\n",
    "print(\"The calculated fitness scores for all individuals in the population are: {}\".format(fitness_scores))\n",
    "print(\"The highest value score ({}) was achieved using the following combination: {}\".format(best_score, best_solution))\n",
    "print(\"This corresponds to the following items: {}\".format(best_solution_items))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3.3 Selection\n",
    "\n",
    "After calculating the fitness of the population, we need to select individuals for reproduction in the next generation. One popular method for doing this is __Tournament Selection__. This method allows us to select individuals with a bias toward stronger solutions while maintaining diversity by occasionally selecting weaker ones. Tournament Selection includes the following steps:\n",
    "\n",
    "1. __Choose a Random Subset__: First, we randomly choose a small subset of individuals from the population. This subset is called the \"tournament.\"\n",
    "2. __Comparison__: Within this subset, we compare the fitness of the individuals. The individual with the highest fitness is the \"winner\" of the tournament.\n",
    "3. __Winner Selection__: The winner is then selected to move on to the next generation, or to participate in reproduction (such as crossover and mutation) to create new solutions.\n",
    "4. __Repeat__: This process is repeated until we have selected enough individuals for the next generation.\n",
    "\n",
    "Here are a couple of reason why tournament selection is a good choice:\n",
    "\n",
    "* __Favors Strong Solutions__: Since the winner is the individual with the highest fitness within the tournament, stronger solutions are more likely to be selected.\n",
    "* __Maintains Diversity__: By randomly selecting the tournament participants, weaker solutions still have a chance to be selected, which helps maintain diversity in the population.\n",
    "* __Adjustable Selection Pressure__: The size of the tournament affects how much we favor stronger solutions:\n",
    "\n",
    "    __(1) Larger tournaments__: (e.g., 5 individuals) make it more likely that a stronger individual will win.\n",
    "  \n",
    "    __(2) Smaller tournaments__: (e.g., 2 individuals) increase the chances of weaker individuals winning, helping to maintain diversity.\n",
    "\n",
    "There are plenty of other selection algorithms of which each has its pros and cons. A quick prompt to the LLM of your choice will help you find out about that. (This time you can write the prompt yourself)\n",
    "\n",
    "Considerig the description of the tournament selection above, we need a function with the following parameters:\n",
    "\n",
    " * __population__: A 2D array representing the population (each row is an individual solution)\n",
    " * __fitness_scores__: A 1D array representing the fitness values of the individuals in the population.\n",
    " * __tournament_size__: The number of individuals to randomly select for each tournament.\n",
    "\n",
    "For simplicty we will select a population that has the same size as the initial population. Additionally, for this function we will not influence the random generators seeds as if we were to do it, the function would always pick the same subset of individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tournament_selection(population: np.ndarray, fitness_scores: np.ndarray, tournament_size: int) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Perform tournament selection to choose individuals for the next generation.\n",
    "\n",
    "    :param population: A 2D array representing the population.\n",
    "    :param fitness_scores: A 1D array representing the fitness values of the individuals in the population.\n",
    "    :param tournament_size: The number of individuals to randomly select for each tournament.\n",
    "    :return: A 2D array of selected individuals (same shape as the input population).\n",
    "    \"\"\"\n",
    "    \n",
    "    # Number of individuals in the population\n",
    "    population_size = population.shape[0]\n",
    "\n",
    "    # Check if tournament size is valid\n",
    "    if tournament_size > population_size:\n",
    "        raise IOError(\"Tournament size must be smaller than the population size.\")\n",
    "    \n",
    "    # Initialize an array to store selected individuals\n",
    "    selected_population = np.zeros_like(population)\n",
    "    \n",
    "    # Loop to select individuals for the next generation\n",
    "    for pos in range(population_size):\n",
    "        \n",
    "        # Randomly select 'tournament_size' individuals from the population\n",
    "        tournament_indices = np.random.choice(population_size, size=tournament_size, replace=False)\n",
    "        \n",
    "        # Get the fitness values for the individuals in the tournament\n",
    "        tournament_fitness = fitness_scores[tournament_indices]\n",
    "        \n",
    "        # Find the index of the individual with the highest fitness in the tournament\n",
    "        winner_index = tournament_indices[np.argmax(tournament_fitness)]\n",
    "        \n",
    "        # Select the winner and add to the new population\n",
    "        selected_population[pos] = population[winner_index, :]\n",
    "    \n",
    "    return selected_population"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With regards to this implementation there are some things that need to be pointed out. Depending on the population size and the selected tournament size:\n",
    "\n",
    "1. The same individual can be picked several times.\n",
    "2. When there are multiple individuals in a tournament that have the same score, the first (with regards to their position in the tournament_fitness array) is always chosen.\n",
    "3. When the tournament size tends to be close to the population size, the propability of picking the same individual in all tournament rounds is high.\n",
    "\n",
    "These effects can mitigated by picking a larger population size and adeqaute tournament sizes. Nevertheless, let's use the function on the  __init_population__ and the corresponding __fitness_scores__. We will use a __tournament_size=2__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The selected population is: \n",
      "[[1 1 1 0 1 0 0 0]\n",
      " [0 1 0 0 0 1 0 0]\n",
      " [1 1 1 0 1 0 0 0]\n",
      " [0 0 1 1 1 1 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 1 1 0 1 0 1 1]\n",
      " [1 0 0 0 0 0 0 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 0 0 0 0 0 0 0]]\n",
      "\n",
      "The initial population was: \n",
      "[[0 1 0 0 0 1 0 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 1 1 0 1 0 1 1]\n",
      " [1 1 1 1 1 1 0 0]\n",
      " [1 1 1 0 1 0 0 0]\n",
      " [0 0 1 1 1 1 1 0]\n",
      " [1 1 0 1 0 1 0 1]\n",
      " [1 0 0 0 0 0 0 0]\n",
      " [0 1 1 0 1 1 1 1]\n",
      " [0 1 0 1 1 1 0 1]]\n",
      "\n",
      "The previous and new populations are equal: False\n"
     ]
    }
   ],
   "source": [
    "# get a new population through tournament selection\n",
    "selected_population = tournament_selection(population=initial_population, fitness_scores=fitness_scores, tournament_size=2)\n",
    "\n",
    "# print the result\n",
    "print(\"The selected population is: \\n{}\".format(selected_population))\n",
    "\n",
    "# comparison to the init population (for visual comparison)\n",
    "print(\"\\nThe initial population was: \\n{}\".format(initial_population))\n",
    "\n",
    "# check programmatically if the two populations are the same\n",
    "populations_are_equal = np.array_equal(initial_population, selected_population)\n",
    "print(\"\\nThe previous and new populations are equal: {}\".format(populations_are_equal))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3.4 Crossover (Recombination)\n",
    "\n",
    "Now that we have selected individuals it is time to perform the Crossover (Recombination) step. Crossover is the process where two parent solutions combine to create one or more offspring solutions. The idea behind crossover is to mix the \"genes\" (bits of information) from the parents, potentially leading to better solutions. \n",
    "\n",
    "Crossover allows the genetic algorithm to explore new regions of the solution space by combining the traits of different individuals. It mimics biological reproduction, where offspring inherit traits from both parents. By mixing the genes from two good solutions, crossover helps to find even better solutions, while also ensuring genetic diversity in the population. This increases the likelihood of evolving solutions that are closer to the optimal. Crossover is constituted of the following steps:\n",
    "\n",
    "1. __Pair Selection__: First, we select two parents from the population.\n",
    "2. __Crossover Point__: A random point, or multiple points, along the binary vector (gene sequence) of each parent is chosen as the crossover point. This point divides the parents into two parts (or more parts).\n",
    "3. __Swapping Genes__: The genes of the two parents are swapped at the crossover point(s). The genes before the crossover point come from the first parent, and the genes after the crossover point come from the second parent\n",
    "\n",
    "In our example we will implement __Single-Point Crossover__. For this, one crossover point is randomly selected along the gene sequence. The part of the genes before this point is copied from one parent, and the part after the point is copied from the other parent. Additionally, we will implement the cross over in such a way, that a pair of parents, always creates a pair of offsprings. Here is a simple example of how this crossover would be performed:\n",
    "\n",
    "    Parent 1: [1, 0, 1, 0, | 1, 1]\n",
    "    Parent 2: [0, 1, 0, 1, | 0, 0]\n",
    "    -------------------------------\n",
    "    Offspring 1: [1, 0, 1, 0, 0, 0] (genes from Parent 1 before the point, Parent 2 after)\n",
    "    Offspring 2: [0, 1, 0, 1, 1, 1] (genes from Parent 2 before the point, Parent 1 after)\n",
    "\n",
    "When performing the crossover step not all pairs of individuals in the selected population are chosen to reproduce. Instead a __crossover probability__ (e.g., 80 %) is chosen to control the crossover rate. This allowsfor some individuals to move on to the next generation without changing their genes, which is in particular useful when an individual is particularly strong.\n",
    "\n",
    "Therefore, we need a function that receives the following inputs:\n",
    "\n",
    "* __population__: A 2D array representing the population.\n",
    "*  __crossover_prob__: The probability with which crossover is performed.\n",
    "*  __seed__: (Optional) seed for the random number generator to ensure reproducibility.\n",
    "\n",
    "Furthermore, for simplicity we will restrict the input of our function to populations that have an even number of individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def population_crossover(population: np.ndarray, crossover_prob: float, seed: Optional[int] = None) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Perform single-point crossover on a randomly paired population, ensuring each individual reproduces once.\n",
    "\n",
    "    :param population: A 2D array representing the population.\n",
    "    :param crossover_prob: The probability with which crossover is performed.\n",
    "    :param seed: Optional seed for the random number generator to ensure reproducibility.\n",
    "    :return: A 2D array representing the new population after crossover.\n",
    "    \"\"\"\n",
    "    \n",
    "    # check if crossover_prob is <= 1 | only numbers between [0, 1] are allowed to express probablity\n",
    "    if not (0 <= crossover_prob <= 1):\n",
    "        raise IOError(\"The provided crossover_prob has to be <=1. Provided crossover_prob: {}\".format(crossover_prob))\n",
    "    \n",
    "    # get the number of individuals in the population and how many \"genes\" each member of the population has\n",
    "    population_size, num_genes = population.shape\n",
    "\n",
    "    # check if the population size is even\n",
    "    if population_size % 2 != 0:\n",
    "        raise IOError(\"The population size must be even to perform crossover. Size of provided population: {}\".format(population_size))\n",
    "\n",
    "    # Set the seed for reproducibility\n",
    "    if seed is not None:\n",
    "        np.random.seed(seed)\n",
    "\n",
    "    # get indicies for a random pairing of the population\n",
    "    pair_indices = _get_random_pair_indices(population_size)\n",
    "\n",
    "    # initialize a new population array to store offspring\n",
    "    new_population = np.zeros_like(population)\n",
    "\n",
    "    # Perform crossover on randomly paired individuals\n",
    "    for pair in pair_indices:\n",
    "\n",
    "        # get the indices of the pairs \n",
    "        idx_parent_1 = pair[0]\n",
    "        idx_parent_2 = pair[1]\n",
    "        \n",
    "        # select individuals (parents) using the shuffled indices\n",
    "        parent_1 = population[idx_parent_1]\n",
    "        parent_2 = population[idx_parent_2]\n",
    "        \n",
    "        # Determine whether to perform crossover based on crossover probability\n",
    "        if np.random.rand() > crossover_prob:\n",
    "            \n",
    "            # If no crossover, copy parents to the next generation unchanged\n",
    "            new_population[idx_parent_1] = parent_1\n",
    "            new_population[idx_parent_2] = parent_2\n",
    "        \n",
    "        else:\n",
    "            # Perform single-point crossover\n",
    "            crossover_point = np.random.randint(1, num_genes)\n",
    "            \n",
    "            # Create offspring by swapping genes at the crossover point\n",
    "            offspring_1 = np.concatenate((parent_1[:crossover_point], parent_2[crossover_point:]))\n",
    "            offspring_2 = np.concatenate((parent_2[:crossover_point], parent_1[crossover_point:]))\n",
    "            \n",
    "            # Add offspring to the new population\n",
    "            new_population[idx_parent_1] = offspring_1\n",
    "            new_population[idx_parent_2] = offspring_2\n",
    "\n",
    "    return new_population\n",
    "\n",
    "\n",
    "def _get_random_pair_indices(population_size: int):\n",
    "    \"\"\"\n",
    "    Generate random pair indices from a population of a given size.\n",
    "    This function shuffles the indices of the population and reshapes them into a \n",
    "    2D array of pairs, where each row contains two indices representing a pair.\n",
    "\n",
    "    :param population_size: The size of the population for which pairs are to be generated.\n",
    "    :returns: A 2D numpy array where each row contains a pair of indices from the population.\n",
    "    \"\"\"\n",
    "\n",
    "    # shuffle the population indices to create random pairs\n",
    "    pair_indices = np.arange(population_size)\n",
    "    np.random.shuffle(pair_indices)\n",
    "\n",
    "    # reshape the indicices array to a 2D array of shape [n, 2]\n",
    "    # as n  is unknown, use -1 for the first dimension so that numpy implicitly infers the dimension\n",
    "    pair_indices = pair_indices.reshape(-1, 2)\n",
    "\n",
    "    return pair_indices"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's use the crossover function on the __selected_population__. We will use a __crossover_prob=0.8 (80%)__. (Try running the crossover function using a crossover_prob=0 and see the result. You will see that in this instance the selected_population and the crossover_population are the same.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The crossover population is: \n",
      "[[1 1 1 0 1 0 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 1 1 0 0 0 0 0]\n",
      " [0 0 1 1 1 1 1 1]\n",
      " [0 1 0 0 0 0 0 0]\n",
      " [0 1 0 0 0 0 0 0]\n",
      " [1 1 1 0 1 0 1 0]\n",
      " [1 0 0 0 1 0 0 0]\n",
      " [0 1 0 0 0 1 0 0]\n",
      " [1 0 0 0 0 0 1 0]]\n",
      "\n",
      "The selected population was: \n",
      "[[1 1 1 0 1 0 0 0]\n",
      " [0 1 0 0 0 1 0 0]\n",
      " [1 1 1 0 1 0 0 0]\n",
      " [0 0 1 1 1 1 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 1 1 0 1 0 1 1]\n",
      " [1 0 0 0 0 0 0 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 0 0 0 0 0 0 0]]\n",
      "\n",
      "The previous and new populations are equal: False\n"
     ]
    }
   ],
   "source": [
    "# perform crossover on population\n",
    "crossover_population = population_crossover(population=selected_population, crossover_prob=0.8, seed=random_seed)\n",
    "\n",
    "# print the result\n",
    "print(\"The crossover population is: \\n{}\".format(crossover_population))\n",
    "\n",
    "# comparison to the init population (for visual comparison)\n",
    "print(\"\\nThe selected population was: \\n{}\".format(selected_population))\n",
    "\n",
    "# check programmatically if the two populations are the same\n",
    "populations_are_equal = np.array_equal(crossover_population, selected_population)\n",
    "print(\"\\nThe previous and new populations are equal: {}\".format(populations_are_equal))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3.5 Mutation\n",
    "\n",
    "In the Mutation Step, we introduce small, random changes to the offspring produced during the crossover step. Mutation is a critical part of the genetic algorithm process because it helps maintain genetic diversity in the population and allows the algorithm to explore new areas of the solution space. \n",
    "\n",
    "Without mutation, the population may become too homogeneous (i.e., too similar to each other) after several generations. This could lead to the algorithm getting stuck in a local optimum, meaning that it finds a solution that appears to be the best in its immediate vicinity but isn't the best possible solution overall. Mutation helps prevent this by introducing random variations that can lead to better solutions being discovered. For mutation the following steps are performed:\n",
    "\n",
    "1. __Probability of Mutation__: Each gene (bit) in an individual's chromosome (binary vector) has a small probability of being flipped (0 → 1 or 1 → 0). This probability is called the mutation rate. Typically, the mutation rate is set to a low value, such as 0.01 or 0.05, meaning that only 1% or 5% of the genes will mutate in a given population.\n",
    "\n",
    "2. __Gene Modification__: For each individual in the population, every bit in the chromosome is tested independently against the mutation rate. If a mutation occurs, that bit is flipped. For example, if a gene that was originally a 0 is mutated, it changes to 1, and vice versa.\n",
    "\n",
    "Let's assume after crossover we have the following individual, where __the first and the fifth gene are mutated__:\n",
    "\n",
    "    Gene Position:       0  1  2  3  4  5  6  7 \n",
    "    ____________________________________________\n",
    "    Individual:         [1, 0, 1, 1, 0, 1, 0, 0]\n",
    "    --------------------------------------------\n",
    "    Mutated Individual: [1, 1, 1, 1, 0, 0, 0, 0]\n",
    "\n",
    "Thus, the function we need to develop needs to have two parameters:\n",
    "\n",
    "* __population__: A 2D array representing the population.\n",
    "* __mutation_rate__: The probability with which mutation is performed on each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_population(population: np.ndarray, mutation_rate: float) -> np.ndarray:\n",
    "    \"\"\"\n",
    "    Apply mutation to a population by randomly flipping bits in each individual's chromosome\n",
    "    with a probability equal to the mutation rate.\n",
    "\n",
    "    :param population: A 2D NumPy array where each row represents an individual in the population (binary vector).\n",
    "    :param mutation_rate: A float representing the probability of mutation for each gene (bit).\n",
    "    :return: A mutated 2D NumPy array representing the new population after mutation.\n",
    "    \"\"\"\n",
    "\n",
    "    # check if mutation_rate is <= 1 | only numbers between [0, 1] are allowed to express probablity\n",
    "    if not (0 <= mutation_rate <= 1):\n",
    "        raise IOError(\"The provided mutation_rate must be <=1. Provided mutation_rate: {}\".format(mutation_rate))\n",
    "    \n",
    "    # get the shape of the population (number of individuals and number of genes)\n",
    "    population_size, num_genes = population.shape\n",
    "    \n",
    "    # create a random matrix for mutation decisions \n",
    "    # (this will generate a matrix of the same size as the population containing values between [0,1])\n",
    "    mutation_matrix = np.random.rand(population_size, num_genes)\n",
    "    \n",
    "    # create a mask where mutation decisions are below the mutation rate\n",
    "    # (the mask is a boolean mask that contains only True and False values. It can be used for indexing)\n",
    "    mutation_mask = mutation_matrix < mutation_rate\n",
    "    \n",
    "    # make a copy of the population\n",
    "    mutated_population = np.copy(population)\n",
    "    \n",
    "    # flip the bits at positions where the mask is True\n",
    "    mutated_population[mutation_mask] = 1 - mutated_population[mutation_mask]\n",
    "    \n",
    "    return mutated_population"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we have done before, we will test the function. We will pass the __crossover_population__ and use a __mutation_rate=0.05__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The mutated population is: \n",
      "[[1 1 1 0 1 0 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [0 1 1 0 0 0 0 0]\n",
      " [0 0 1 1 1 0 1 1]\n",
      " [0 1 0 0 0 0 0 0]\n",
      " [0 1 0 0 0 1 0 0]\n",
      " [1 1 1 0 1 0 1 0]\n",
      " [1 0 0 1 1 0 0 0]\n",
      " [0 1 0 0 0 1 0 0]\n",
      " [1 0 0 0 0 0 1 0]]\n",
      "\n",
      "The crossover population was: \n",
      "[[1 1 1 0 1 0 1 0]\n",
      " [0 1 0 0 0 0 1 0]\n",
      " [1 1 1 0 0 0 0 0]\n",
      " [0 0 1 1 1 1 1 1]\n",
      " [0 1 0 0 0 0 0 0]\n",
      " [0 1 0 0 0 0 0 0]\n",
      " [1 1 1 0 1 0 1 0]\n",
      " [1 0 0 0 1 0 0 0]\n",
      " [0 1 0 0 0 1 0 0]\n",
      " [1 0 0 0 0 0 1 0]]\n",
      "\n",
      "The previous and new populations are equal: False\n"
     ]
    }
   ],
   "source": [
    "# apply mutation of genes to the population\n",
    "mutated_population = mutate_population(population=crossover_population, mutation_rate=0.05)\n",
    "\n",
    "# print the result\n",
    "print(\"The mutated population is: \\n{}\".format(mutated_population))\n",
    "\n",
    "# comparison to the init population (for visual comparison)\n",
    "print(\"\\nThe crossover population was: \\n{}\".format(crossover_population))\n",
    "\n",
    "# check programmatically if the two populations are the same\n",
    "populations_are_equal = np.array_equal(crossover_population, mutated_population)\n",
    "print(\"\\nThe previous and new populations are equal: {}\".format(populations_are_equal))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2.3.6 Termination\n",
    "\n",
    "The code that we have written and executed above consitutes one iteration of a genetic algorithm. Each iteration is also called __\"generation\"__ as what we did was starting with an __initial population__ and obtained a __mutated population__/__new_population__. A genetic algorithm runs for several generations until a stopping criterion is met. Below you can find some suitable stopping criteria:\n",
    "\n",
    "* __Number of Generations__: The algorithm stops after a fixed number of generations. This is a straightforward approach and ensures that the algorithm does not run indefinitely.\n",
    "* __Convergence__: The algorithm stops when the population converges, meaning there is little or no improvement in the best fitness score over a specified number of generations. This indicates that the algorithm has potentially found a good solution and further iterations may not yield significant improvements.\n",
    "* __Target Fitness__: The algorithm stops when an individual with a fitness score that meets or exceeds a predefined target fitness level is found. This criterion ensures that the algorithm stops once an acceptable solution is discovered.\n",
    "* __Diversity Loss__: The algorithm stops when the diversity of the population falls below a certain threshold, meaning that the population has become too homogeneous and lacks variability.\n",
    "* __Elapsed Time__: The algorithm stops after a predefined amount of time has passed. This can be useful when computational resources are limited or when a solution needs to be found within a specific time frame.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Exercises\n",
    "\n",
    "__Exercise 1__:\n",
    "As you now know how each step is implemented, it is time for you to put everything together into an overall function that incorporates the functions from above. You will implement this function using the __number of generations__ as convergence criterion.\n",
    "\n",
    "__Tasks__:\n",
    "\n",
    "* Implement a function that runs the entire GA for the Knapsack Problem\n",
    "* The function should have the following parameters: \n",
    "     >* __knapsack_properties__: Dictionary containing the properties of the knapsack (item weights, item values, item names, max capacity).\n",
    "     >* __population_size__: The number of individuals in the population.\n",
    "     >*__tournament_size__: The number of individuals to randomly select for each tournament.\n",
    "     >* __crossover_prob__: The probability with which crossover is performed.\n",
    "     >* __mutation_rate__:  A float in the range [0, 1] representing the probability of mutation for each gene (bit).\n",
    "     >* __num_generations__: The number of generations for which the genetic algorithm should run\n",
    "     >* __seed__: (Optional) seed for the random number generator to ensure reproducibility.\n",
    "\n",
    " \n",
    "* The function should output the following:\n",
    "    >* __best_fitness_scores__: A List containing the best fitness score for each generation.\n",
    "    >* __best_items__: A List containing the items that achieved the highest fitness score over all generations.\n",
    "\n",
    "* The function should be called __knapsack_genetic_algorithm__.\n",
    "* Don't forget to add a __docstring__, __typing hints__, and __comments__.\n",
    "\n",
    "The dictionary containing the knapsack properties is defined below:\n",
    "\n",
    "__Hints__:\n",
    "\n",
    "* Hint 1: For storing the best_fitness_scores and best_items you could use lists that you can update in each generation if it is needed. (The best_items only need to be updated if the fitness score of the current generation is better than the previous generation's score)\n",
    "* Hint 2: Some of the code written above might help you in getting the best_fitness_scores for each generation and finding the best_items\n",
    "corresponding to that score. We suggest wrapping that code into its own function.\n",
    "* Hint 3: Whenever you are writing code, try writing it as memomry efficient as possible. Contrary to what we did above, always storing the \"evolved population\" (initial_population, selected_population, etc.) into a new variable is not recommended. Instead one variabale that holds the population (and is overwritten when needed) should be enough.\n",
    "* Hint 4: Given that the knapsack properties dictionary contains multiple types, the typing hint for it is not obvious. You might want to consider asking the LLM of your choice for some help.\n",
    "* Hint 5 (Optional): Use the [__tqdm__](https://www.askpython.com/python-modules/python-progress-bars-with-tqdm) module to display a progress bar when running the for loop that runs over the number of generations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# defining knapsack dictionary keys as constants \n",
    "# (these are defined as constants as they will stay static throughout the entire runtime of the code)\n",
    "ITEM_WEIGHTS = 'item_weights'\n",
    "ITEM_VALUES = 'item_values'\n",
    "ITEM_NAMES = 'item_names'\n",
    "MAX_CAPACITY = 'max_capacity'\n",
    "\n",
    "# definition of the knapsack properties as dictionary\n",
    "knapsack_properties = {ITEM_WEIGHTS: np.array([5.0, 2.5, 3.5, 1.0, 1.5, 0.5, 4.0, 1.5]), \n",
    "                       ITEM_VALUES: np.array([60, 40, 30, 10, 20, 15, 50, 25]),\n",
    "                       ITEM_NAMES: [\"tent\", \"sleeping bag\", \"camping stove\", \"water bottle\", \"first aid kit\", \"flashlight\", \"food supplies\", \"jacket\"], \n",
    "                       MAX_CAPACITY: 10}\n",
    "\n",
    "# reminder: to access the needed property from the dictionary you need to use the corresponding key\n",
    "# e.g., item_weights = knapsack_properties[ITEM_WEIGHTS]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 2__:\n",
    "Now it is time to test your function.\n",
    "\n",
    "__Tasks__:\n",
    "\n",
    "* Run the the function with different parameters to investigate how the output changes. Which is the highest score you can achieve?\n",
    "* Print the highest fitness score (utility points) and the best items.\n",
    "* Plot the best_fitness_scores to show how the scores evolve over the generations. Add approriate labeling for the x- and y-axis, and add a title."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Optimizing Filter Parameters for Noise Removal in ECG Signals\n",
    "\n",
    "The Electrocardiogram (ECG) is a recording of the electrical activity of the heart over time. It is widely used in medical settings to monitor heart health and detect abnormalities. The ECG signal consists of several key components, including the P-wave, QRS complex, and T-wave, which correspond to different phases of the heart's electrical cycle.\n",
    "\n",
    "An ideal ECG signal is clean and exhibits clear peaks and valleys. However, in practice, ECG signals often contain noise that can distort these key features, __making it difficult to interpret the signal accurately__.\n",
    "\n",
    "The example we are going to explore today is relatively simple as we are only going to tackle one particular noise source. However, in real-world scenarios you often encounter a superposition of noise types which makes it hard to filter these out with generic filter pipelines. Efficient noise removal is still an open research question. Here is an example of a recent publication from our Lab ([LIBPhys Biosignals](https://www.libphysbiosi.gnals.org/)): [Cleaning ECG with Deep Learning: A Denoiser Tested in Industrial Settings](https://link.springer.com/article/10.1007/s42979-024-03017-7?utm_source=rct_congratemailt&utm_medium=email&utm_campaign=oa_20240703&utm_content=10.1007%2Fs42979-024-03017-7).\n",
    "\n",
    "### 3.1 Common Noise Sources in ECG Signals\n",
    "\n",
    "Noise sources that can be commonly found in ECG signals are:\n",
    "\n",
    "* __Muscle Activity (EMG Noise)__: Electrical activity from nearby muscles can cause high-frequency noise in the ECG signal.\n",
    "* __Power Line Interference__: Electrical devices around the patient can introduce a 50/60 Hz interference into the ECG.\n",
    "* __Motion Artifacts__: Movements of the patient, such as shifting or breathing, can create abrupt changes in the signal.\n",
    "* __Baseline Wander__: A slow drift in the baseline of the ECG signal, often caused by breathing or body movements, leading to low-frequency noise.\n",
    "\n",
    "For this example we will focus on __baseline wander__ noise, which is a type of __low-frequency noise__ commonly observed in ECG recordings that leads to a slow, drifting oscillation of the ECG signal baseline. While baseline wander does not significantly distort the higher-frequency components of the signal (like the QRS complex), it can make it difficult to interpret the overall shape of the ECG and measure the key intervals accurately. As baseline wander is a low-frequency noise it can be removed using a __high-pass filter__.\n",
    "\n",
    "### 3.2 High-Pass Fitlers and Their Role in ECG Signal Processing\n",
    "\n",
    "A high-pass filter is a type of filter that allows high-frequency components of a signal to pass through while attenuating (reducing) the amplitude of lower-frequency components. In the context of ECG signal processing, high-pass filters are used to remove low-frequency noise while retaining the important higher-frequency features of the signal, such as the QRS complex. \n",
    "\n",
    "A high-pass filter has two parameters:\n",
    "\n",
    "* __Cutoff Frequency__: The key parameter of a high-pass filter is the cutoff frequency. Frequencies above this threshold are allowed to pass, while frequencies below it are attenuated.\n",
    "\n",
    "* __Filter Order__: The filter order controls how sharp the cutoff is. A higher filter order results in a steeper transition from the passband (high frequencies) to the stopband (low frequencies). However, higher orders can introduce phase distortion, so a balance must be found between a steep cutoff and minimal signal distortion.\n",
    "\n",
    "Our goal is to __optimize__ the filter’s parameters, __cutoff frequency__ and __filter order__, using a genetic algorithm. This will ensure that baseline wander is minimized while preserving the essential features of the ECG signal.\n",
    "\n",
    "### 3.3 Signal-to-Noise Ration (SNR)\n",
    "\n",
    "To evaluate the effectiveness of the filter, we will measure the __Signal-to-Noise Ratio (SNR)__, compares the level of the desired signal to the level of the noise. SNR is a widely used metric in signal processing to quantify the clarity of the signal after filtering. A high SNR indicates, that a signal is of good quality (has low noise).\n",
    "\n",
    "The SNR, measured in __dB (decibel)__ can be calculated as follows:\n",
    "\n",
    "$$\n",
    "\\text{SNR} = 10 \\cdot \\log_{10} \\left( \\frac{\\text{Signal Power}}{\\text{Noise Power}} \\right)\n",
    "$$\n",
    "\n",
    "With:\n",
    "\n",
    "$$\n",
    "\\text{Signal Power} = \\frac{1}{N} \\sum_{i=1}^{N} x[i]^2,\n",
    "$$\n",
    "\n",
    "$$\n",
    "\\text{Noise Power} = \\frac{1}{N} \\sum_{i=1}^{N} \\left( \\hat{x}[i] - x[i] \\right)^2\n",
    "$$\n",
    "\n",
    "Where: \n",
    "- \\\\( N \\\\) is the total number of samples in the signal.\n",
    "- \\\\( x[i] \\\\) is the amplitude of the clean signal at the \\(i\\)-th sample.\n",
    "- \\\\( \\hat{x}[i] \\\\) is the filtered (noisy) signal.\n",
    "\n",
    "In this example, the GA will use __SNR as the fitness function__. The filter parameters will be optimized to maximize the SNR, ensuring that the filtered ECG signal has minimal noise and retains its important features."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Exercises\n",
    "\n",
    "Now it is up to you to write a genetic algorithm from scratch. Just to re-iterate, your goal will be:\n",
    "\n",
    "* __Parameters to Optimize__: You will be using a GA to optimize the __cutoff frequency__ and __filter order__ to remove baseline wander noise from a ECG signal.\n",
    "* __Fitness Function__: As the fitness function you will use the __SNR__. A __high SNR__ indicates that the signal is of __good quality__.\n",
    "\n",
    "But before we start writing the GA let's get familiarized with the signals and write some functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports (only importing additional modules that are needed all other modules were imported above)\n",
    "from scipy import signal # this module will be needed for the high-pass filter\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 1__:\n",
    "Get yourself acquainted with the signals that we are working with. In the Data folder you can find two numpy files containing the same signal.\n",
    "\n",
    "* __clean_ECG.npy__: contains the clean (noise-free) ECG signal\n",
    "* __noisy_ECG.npy__: contains the same signal, but corrupted by __baseline wander__\n",
    "\n",
    "Both signals were acquired with a sampling frequency of $f_s=360 Hz$.\n",
    "\n",
    "__Tasks__:\n",
    "\n",
    "* Define the sampling frequency as a constant (as it will remain static throughout the entire code runtime).\n",
    "* Write a function that plots the two signals into the same plot:\n",
    "    >* Call the function __plot_ECG_signal_comparison__.\n",
    "    >* The function should receive as input the clean and noisy signal.\n",
    "    >* Add approriate labeling for the x- and y-axis, a title, and a legend. The signals should be named accordingly in the legend.\n",
    "* Load both signals.\n",
    "* Plot the signals usig your function\n",
    "\n",
    "\n",
    "__Hints__:\n",
    "* Hint 1: Use the [numpy.load()](https://numpy.org/doc/stable/reference/generated/numpy.load.html) function to load the data\n",
    "* Hint 2: Use [matplotlib.pyplot.plot()](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot.html) for plotting the data.\n",
    "* Hint 2: The [matplotlib.pyplot.plot()](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot.html) function has a parameter that allows for definiton of a legend label."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 2__:\n",
    "Write a function that calculates the SNR as defined above.\n",
    "\n",
    "__Tasks__:\n",
    "* The function should have as input:\n",
    "    >* __clean_ecg__: the clean ECG signal. (Type: np.array)\n",
    "    >* __filtered_ecg__: the filtered ECG signal (Type: np.array)\n",
    "* The function should output:\n",
    "    >* __snr__: the SNR rounded to the second digit after the comma.\n",
    "* Call the function __calc_snr__.\n",
    "* Calculate the SNR between the original clean and noisy signal. (you can pass the noisy signal as the \"filtered_signal\".)\n",
    "* Print the result.\n",
    "\n",
    "__Hints__:\n",
    "* Have a good look at the formulas defined above, some terms cancel each other out.\n",
    "* There is a case for which the calculation of the SNR is not possible. Add a ValueError exception for this case.\n",
    "* The python __math__ module has a [math.log()](https://docs.python.org/3/library/math.html) function that you can use.\n",
    "* The numpy module has a [numpy.sum()](https://numpy.org/doc/stable/reference/generated/numpy.sum.html) function.\n",
    "* The numpy module has a [numpy.round()](https://numpy.org/doc/stable//reference/generated/numpy.round.html) function.\n",
    "* To square a value in python you simply have to use the ** operator. (e.g.: x_squared = x**2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 3__:\n",
    "Write a function that performs high-pass filtering on the signal.\n",
    "\n",
    "__Tasks__:\n",
    "* Call the function __butter_highpass_filter__\n",
    "* The function should receive the following inputs:\n",
    "    >* __ecg_signal__: 1D numpy array containing the signal that is supposed to be filtered.\n",
    "    >* __cutoff_freq__: The cutoff frequency of the high-pass filter. (Type: float)\n",
    "    >* __fs__: The signals sampling rate. (Type: int)\n",
    "    >* __order__: The order of the filter. (Type: int)\n",
    "    \n",
    "* the function should output\n",
    "    >* __filtered_ecg_signal__: The filtered ECG signal.\n",
    "\n",
    "__Hints__:\n",
    "* Check out the documentation for the [scipy.signal.butter](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.butter.html) function.\n",
    "* As the GA will also explore higher order filters, use the second-order sections (sos) output format (see at the bottom of the documentation page)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 4__:\n",
    "Now you will start implementing the steps of the GA to optimize the filter. First you will start with the population initialization function. Be aware that this time we do not have a binary problem. Instead we have a continuous (for the cutoff frequency) and a discrete (for the filter order) problem as the parameters for the high-pass filter can assume continuous/discrete values. As we already know that basline wander is a low frequency noise, the GA does not have to explore the full search space. Thus, we will limit the values we will explore to the following:\n",
    "\n",
    "* __cutoff frequency (\\\\(f_c \\\\))__: \\\\( f_c \\in [0.05 Hz, 1 Hz)  \\\\) and \\\\(f_c \\in \\mathbb{R} \\\\) (floating point numbers)\n",
    "* __order (odr)__: \\\\( odr\\in [1, 5] \\\\) and \\\\( odr \\in \\mathbb{N} \\\\) (integer numbers)\n",
    "\n",
    "\n",
    "Note: The limitation of the cutoff frequency to a maximum of 1 Hz is done as the relevant information of the ECG lies between \\\\( [1 Hz, 45 Hz] \\\\)\n",
    "\n",
    "__Tasks__:\n",
    "* Define the cutoff frequency and order intervals as constants.\n",
    "* Write the function for the population initialization.\n",
    "* call the function __init_population__\n",
    "* The function should receive the following inputs:\n",
    "    >* __population_size__: The size of the population. (Type: int)\n",
    "    >* __fc_range__: The range that should be explored for the cutoff frequencies. (Type: Tuple[float, float])\n",
    "    >* __order_range__: The range that should be explored for the filter order. (Type: Tuple[int, int])\n",
    "    >* __seed__: seed for the random number generator to ensure reproducibility.\n",
    "\n",
    "* The function should output:\n",
    "    >* __population__: A list of tuples containing the generated population. (Type: List[Tuple[float, int]])\n",
    "\n",
    "* Test your function by running it with a population_size=10 and print the output\n",
    "\n",
    "__Hints__:\n",
    "* Hint 1: You can define the constants for the cutoff frequency and order intervals as tuples. \n",
    "* Hint 2: As the parameters you are optimizing have different types (float and int), it is not recommended to store the population into a np.array. We recommend to use a list instead.\n",
    "* Hint 3: For generating random floats within an interval you should have a look at the [numpy.random.unifrom()](https://numpy.org/doc/stable/reference/random/generated/numpy.random.uniform.html) function. As the function returns numbers with a high precision (several digits after the comma), you should round the output the the second digit after the comma.\n",
    "* Hint 4: For generating random integers within an interval you should have a look at the [numpy.random.randint()](https://numpy.org/doc/stable/reference/random/generated/numpy.random.uniform.html) function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 5__:\n",
    "Now you will write the function that calculates the fitness scores for each individual within your population. Remeber, the fitness is claculated this time based on the SNR between the clean and the filtered signal. Thus, you will have to use the functions you implemented in __Exercise 2__ and __Exercise 3__.\n",
    "\n",
    "__Tasks__:\n",
    "* Call the function __calculate_fitness__.\n",
    "* The function should receive the following inputs:\n",
    "    >* __population__: A list of tuples containing the population. (Type: List[Tuple[float, int]])\n",
    "    >* __clean_ecg__: The clean ECG signal. (Type: np.array)\n",
    "    >* __noisy_ecg__: The noisy ECG signal. (Type: np.array)\n",
    "    >* __fs__: The signals sampling rate. (Type: int)\n",
    " \n",
    "* The function should output:\n",
    "    >* __fitness_scores__: A list containing the fitness scores for the individuas contained in population. (Type: List[float])\n",
    "* Test your function by running it with the initial_population you generated in the previous exercise and print out the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 6__: Now that you have the fitness scores, you can implement the selection step. You will again use __tournament selection__. The tournament selection function we implemented for the knapsack problem can not be used in this case as it uses numpy.arrays for the calculation. This time you have to implement it using the list objects you are using. However, the logic (i.e., pefromed steps) are the same, so you can use the function as a reference.\n",
    "\n",
    "__Tasks__:\n",
    "* Implement the tournament selection, adapting it to the new data types you are using.\n",
    "* Call the function __tournament_selection__.\n",
    "* The function should receive the following inputs:\n",
    "    >* __population__: A list of tuples containing the population. (Type: List[Tuple[float, int]])\n",
    "    >* __fitness_scores__: A list containing the fitness scores for the individuas contained in population. (Type: List[float])\n",
    "    >* __tournament_size__: The number of individuals to randomly select for each tournament. (Type: int)\n",
    "* The function should output:\n",
    "    >* __selected_population__: A list of tuples containing the selected population. (Type: List[Tuple[float, int]])\n",
    "* Test your function using a __tournament_size=2__ and print the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 7__: Next, you need to implement the crossover function. The crossover function in this case will be a bit simpler than in the Knapsack Problem example as each individual just consists of two elements. You basically just have to swap out the second element of the Tuple representing a parent.\n",
    "\n",
    "__Example__: Let's asume we have the following two parents and their offspring: \n",
    "\n",
    "    Parent 1: (0.74, 2)\n",
    "    Parent 2: (0.20, 3)\n",
    "    -------------------\n",
    "    offspring 1: (0.74, 3)\n",
    "    offspring 2: (0.20, 2)\n",
    "\n",
    "__Tasks__:\n",
    "* Implement the crossover function.\n",
    "* Call the function __crossover__.\n",
    "* The function should receive the following inputs:\n",
    "    >* __population__: A list of tuples containing the population. (Type: List[Tuple[float, int]])\n",
    "    >* __crossover_prob__: The probability with which crossover is performed. (Type: float)\n",
    "    >* __seed__: seed for the random number generator to ensure reproducibility. (Type: int, Optional)\n",
    "* The function should return:\n",
    "    >*  __crossover_population__: A list of tuples containing the crossover population. (Type: List[Tuple[float, int]])\n",
    "* Test your function using a __crossover_prob=0.8__ and print the output.\n",
    "\n",
    "__Hints__:\n",
    "* Hint 1: The function __get_random_pair_indices()__ that we implemented above can be reused in this example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 8__: Almost done! Now you have to implement the mutation function. As we are working with continuous and discrete values, the mutation approach we used in the knapsack problem is not adequate. Instead you will implment another approach called __Creep Mutation__. \n",
    "\n",
    "__Creep Mutation__: Is a mutation technique where small, incremental changes are applied to a gene's value rather than large, random mutations. It is typically used when the genes represent real numbers or continuous variables, and the mutation is intended to explore nearby regions of the solution space.\n",
    "\n",
    "In our case we will increse or decrease the variables by adding values from the following ranges\n",
    "* __cutoff frequency creep (\\\\( f_{c,cr} \\\\))__: \\\\( f_{c,cr} \\in [-0.1 Hz, 0.1 Hz]  \\\\) and \\\\(f_{c,cr} \\in \\mathbb{R} \\\\)\n",
    "* __order creep (\\\\( odr_{cr} \\\\))__: \\\\( odr_{cr}\\in [-1, 1] \\\\) and \\\\( odr_{cr} \\in \\mathbb{N} \\\\)\n",
    "\n",
    "In the case that the addition of the value taken out of the creep range, results in a value that is outside of the range that we are exploring for the frequency cutoff and the filter order, you will just clip it to the maximum or minimum of the range.\n",
    "\n",
    "__Example__: Let's have a look at two examples:\n",
    "\n",
    "    order : 5\n",
    "    value taken out of the order creep range: 1\n",
    "    ____________________________________________________\n",
    "    mutated value: 5 + 1 = 6  --> out of range clip: to 5 (defined range for order: [1, 5])\n",
    "\n",
    "    frequency cutoff : 0.1 Hz\n",
    "    value taken out of the frequency creep range: -0.1 Hz\n",
    "    ____________________________________________________\n",
    "    mutated value: 0.1 Hz + (-0.1) Hz = 0 Hz  --> out of range clip: to 0.05 Hz (defined range for frequency cutoff: [0.05, 1] Hz)\n",
    "\n",
    "__Tasks__:\n",
    "* Define the minimum and maximum of both ranges as constants.\n",
    "* Write the function for mutating the individuals in the population.\n",
    "* Call the function __mutate_population__\n",
    "* The function should have the following parameters:\n",
    "    >* __population__: A list of tuples containing the population. (Type: List[Tuple[float, int]])\n",
    "    >* __mutation_rate__: The probability with which mutation is performed on each gene. (Type: float)\n",
    "    >* __cutoff_range__: The allowed range for the cutoff frequency (min, max). (Type: Tuple[float, float])\n",
    "    >* __order_range__: The allowed range for the filter order (min, max). (Type: Tuple[float, float])\n",
    "* The function should return:\n",
    "    >* __mutated_population__:  A list of tuples containing the population. (Type: List[Tuple[float, int]])\n",
    "* Test your function using a __mutation_rate=0.1__ and print the result.\n",
    "\n",
    "__Hints__:\n",
    "* For selecting creep values for the cutoff frequency, you can use [numpy.random.unifrom()](https://numpy.org/doc/stable/reference/random/generated/numpy.random.uniform.html) with the specified range. Round the creep value for the cutoff frequency to two digits after the comman.\n",
    "* For selecting creep values for the filter order, you can use [numpy.random.choice()](https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html) and pass the range as a list. No other parameters are needed.\n",
    "* For cliping values to a certain range you can use [numpy.clip()](https://numpy.org/doc/stable/reference/generated/numpy.clip.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Exercise 9__: Final exercise! You made it!! Now it is your task to put everything together. This time you don't need to write a function. Just a code that combines everything you did above.\n",
    "\n",
    "__Tasks__:\n",
    "* Combine the code to run the the genetic algorithm.\n",
    "* Use the __Number of Iterations__ as termination criterion.\n",
    "* Get the best fitness score of each iteration and save it into a list.\n",
    "* Get the best filter parameters over all generation.\n",
    "* Plot the best fitness score over the number of generations.\n",
    "* Apply the filter with the best parameters to the noisy signal.\n",
    "* Plot both signals (the clean and the filtered signal) together into one plot.\n",
    "\n",
    "__Hints__:\n",
    "* Hint 1: It is beneficial to define all your adjustable parameters at the top of the code, so that you have them all together in one place to change them more effectively.\n",
    "* Hint 2: Use the [__tqdm__](https://www.askpython.com/python-modules/python-progress-bars-with-tqdm) module to display a progress bar when running the for loop that runs over the number of generations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "vscode": {
   "interpreter": {
    "hash": "ee3b8b0f13ce1b575702cf6f1b3dd3d8df18dc5f202e0bca1f4b2f664d388be6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
